src/pybind/mgr/dashboard/controllers/rbd.py

```diff
@@ -1,136 +1,431 @@
 # -*- coding: utf-8 -*-
+# pylint: disable=too-many-arguments,too-many-locals,unused-argument,
+# pylint: disable=too-many-statements,too-many-branches
 from __future__ import absolute_import
 
 import math
 import cherrypy
+import rados
 import rbd
 
-from . import ApiController, AuthRequired, RESTController
+from . import ApiController, AuthRequired, RESTController, Task
 from .. import mgr
+from ..services.ceph_service import CephService
 from ..tools import ViewCache
 
 
-@ApiController('rbd')
+# pylint: disable=inconsistent-return-statements
+def _rbd_exception_handler(ex):
+    if isinstance(ex, rbd.OSError):
+        return {'status': 409, 'detail': str(ex), 'errno': ex.errno,
+                'component': 'rbd'}
+    elif isinstance(ex, rados.OSError):
+        return {'status': 409, 'detail': str(ex), 'errno': ex.errno,
+                'component': 'rados'}
+    raise ex
+
+
+def RbdTask(name, metadata, wait_for):
+    return Task("rbd/{}".format(name), metadata, wait_for,
+                _rbd_exception_handler)
+
+
+def _rbd_call(pool_name, func, *args, **kwargs):
+    with mgr.rados.open_ioctx(pool_name) as ioctx:
+        func(ioctx, *args, **kwargs)
+
+
+def _rbd_image_call(pool_name, image_name, func, *args, **kwargs):
+    def _ioctx_func(ioctx, image_name, func, *args, **kwargs):
+        with rbd.Image(ioctx, image_name) as img:
+            func(ioctx, img, *args, **kwargs)
+
+    return _rbd_call(pool_name, _ioctx_func, image_name, func, *args, **kwargs)
+
+
+RBD_FEATURES_NAME_MAPPING = {
```

Comment:
How can we make sure, this is not getting out-of-date?

Reply:
don't know, we should figure out a better way in the future.